### PR TITLE
feat(hydrator): add commit-server component

### PR DESCRIPTION
**Pull request by crenshaw-dev**
I'm breaking the hydrator into separate PRs. This is party to make them easier to review, partly because the commit history on the original PR is pretty messy at this point.

This PR is just for the new commit-server component. In this code, the commit server isn't invoked by anything besides tests, and the k8s manifests are not included in the rendered install manifests.

I've added everyone who helped as a co-author on this PR.

### CODE REVIEW

File: commitserver/commit/secure_mkdir_default.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+package commit
+
+import (
+	"fmt"
+	"os"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+)
+
+func SecureMkdirAll(root, unsafePath string, mode os.FileMode) (string, error) {
```
Comment (leoluz) [inline commitserver/commit/secure_mkdir_default.go:12]: It seems an useful function that can be used in various contexts. Have you considered declaring it in `util/io/files` package?

File: commitserver/commit/secure_mkdir_linux.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+package commit
+
+import (
+	"fmt"
+	"os"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+)
+
+func SecureMkdirAll(root, unsafePath string, mode os.FileMode) (string, error) {
```
Comment (leoluz) [inline commitserver/commit/secure_mkdir_linux.go:12]: Have you considered declaring it in util/io/files package?

File: commitserver/commit/secure_mkdir_default.go (hunk 1 above)
Comment (leoluz) [reply commitserver/commit/secure_mkdir_default.go:12]: also, please add godoc

File: util/git/client.go [hunk 3]
```diff
@@ -80,11 +80,23 @@ type Client interface {
 	IsAnnotatedTag(string) bool
 	ChangedFiles(revision string, targetRevision string) ([]string, error)
 	IsRevisionPresent(revision string) bool
+	// SetAuthor sets the author name and email in the git configuration.
+	SetAuthor(name, email string) (string, error)
+	// CheckoutOrOrphan checks out the sync branch. If the branch does not exist, it creates an orphan branch.
```
Comment (leoluz) [inline util/git/client.go:85]: You are documenting with as "sync branch" but this function can be used in different contexts than hydrator.

File: util/git/client.go [hunk 4]
```diff
@@ -80,11 +80,23 @@ type Client interface {
...
+	CheckoutOrOrphan(branch string, submoduleEnabled bool) (string, error)
+	// CheckoutOrNew checks out the target branch. If the branch does not exist, it creates an empty branch based
+	// on the sync branch.
```
Comment (leoluz) [inline util/git/client.go:88]: Same here. Please describe the function in general terms instead of using hydrator terms as sync branch.

**Review by leoluz — CHANGES_REQUESTED** [review]
Please check my first pass comments

File: commitserver/apiclient/clientset.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package apiclient
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply commitserver/apiclient/clientset.go:1]: Correct!

File: commitserver/commit/commit.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	Commands []string `json:"commands"`
+}
+
+// TODO: make this configurable via ConfigMap.
+var manifestHydrationReadmeTemplate = `
+# Manifest Hydration
+
+To hydrate the manifests in this repository, run the following commands:
+
+` + "```shell\n" + `
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply commitserver/commit/commit.go:182]: We definitely need to standardize our templating languages more. But fasttemplate is unmaintained, so I think we should standardize on go template.

File: commitserver/commit/secure_mkdir_default.go (hunk 1 above)
Comment (crenshaw-dev) [reply commitserver/commit/secure_mkdir_default.go:12]: Yep, good idea to move it. I looked around for other use cases and haven't seen any yet. But util/io/files is a better home.

File: commitserver/commit/commit.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		HydratedRevision: sha,
+	}, nil
+}
+
+// handleCommitRequest handles the commit request. It clones the repository, checks out the sync branch, checks out the
+// target branch, clears the repository contents, writes the manifests to the repository, commits the changes, and pushes
+// the changes. It returns the output of the git commands and an error if one occurred.
+func (s *Service) handleCommitRequest(ctx context.Context, logCtx *log.Entry, r *apiclient.CommitHydratedManifestsRequest) (string, string, error) {
+	logCtx.Debug("Initiating git client")
+	gitClient, dirPath, cleanup, err := s.initGitClient(ctx, logCtx, r)
```
[the remark this answers is not in the document]
Comment (leoluz) [reply commitserver/commit/commit.go:64]: Sure. I agree with the local cache and I wasn't suggesting an early optimization. My point is specifically on the current fetch/clone command as it seems that we are bringing the whole repo. Would it be possible to do something like a shallow clone and bring just the necessary hash?

[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply commitserver/commit/commit.go:64]: That's a good point... let me look at the possibility of a shallow clone.

If it's a lot of work, I suggest we skip it. Shallow clones and sparse checkouts make sense for use cases where you clone every time (like the current commit server implementation) but don't really make sense when you have a shared clone. If we switch to persistent clones, I think the shallow/sparse logic would become useless.

**Comment by wanghong230** [comment]
Still TBD for me to see whether this big feature will go straight into the main project here.

**Comment by crenshaw-dev** [comment]
@wanghong230 understood. Let's talk at the contributors meeting tomorrow. I have a vacation coming up,  so I'm trying to get thongs in good shape for 2.13 assuming we go that direction. If we merge anything, it can be reverted before 2.13 rc1.

**Comment by crenshaw-dev** [comment]
@leoluz I'm going to close this in favor of one based on a branch in this repo instead of my fork. That'll let me line up sequential PRs and have the diffs reviewable here.

**Comment by crenshaw-dev** [comment]
Closing in favor of https://github.com/argoproj/argo-cd/pull/19537

**Pull request closed without merging**